### PR TITLE
core: allow illegal header keys to be ignored #3133

### PR DESCRIPTION
**Pull request by nolledge**
* reuse ignore illegal header value configuration
* configuration now applies to both, header and key
* BREAKING: name of configuration key changed

https://github.com/akka/akka-http/issues/3133

References #3133

### CODE REVIEW

**Comment by lightbend-cla-validator** [comment]
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

**Comment by lightbend-cla-validator** [comment]
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

**Comment by lightbend-cla-validator** [comment]
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

**Comment by lightbend-cla-validator** [comment]
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

File: akka-http-core/src/main/resources/reference.conf [hunk 1]
```diff
@@ -664,6 +664,15 @@ akka.http {
     # `full`  : the full error details (potentially spanning several lines) are logged
     error-logging-verbosity = full
 
+    # Configures the processing mode when encountering illegal characters in
+    # header name of response.
+    #
+    # Supported mode:
+    # `error`  : default mode, throw an ParsingException and terminate the processing
```
Comment (raboof) [inline akka-http-core/src/main/resources/reference.conf:671]:
```suggestion
    # `error`  : default mode, reject the request
```

**Review by raboof — COMMENTED** [review]
This is looking quite good!

3 more things:
* We should double-check this change doesn't affect performance. It looks like it's probably fine, but could you run the `HeaderParserBenchmark` to make sure?
* The commit has author `Christof Nolle <mail@christof-nolle.de>`, which is not tied to your github account. Could you add that email to your github account, or update the commit to use an email address that is known in github for you?
* The tool we use for checking binary compatibility, MiMa, is producing warnings. I think they are all about internal classes (for which we don't promise binary compatibility, like `ParserSettingsImpl`), or about added methods to public classes that are not for user extension (like `ParserSettings`) - so I think they can be ignored. Could you add a mima backwards-excludes file with those exclusions?

**Comment by lightbend-cla-validator** [comment]
Hi @nolledge,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[https://www.lightbend.com/contribute/cla](https://www.lightbend.com/contribute/cla)

**Comment by nolledge** [comment]
Regarding the three open points:

1. I ran the HeaderParserBenchmark before and after my changes. I attached the results here:

[header_benchmark_after.txt](https://github.com/akka/akka-http/files/5604858/header_benchmark_after.txt)
[header_benchmark_before.txt](https://github.com/akka/akka-http/files/5604859/header_benchmark_before.txt)

2. I linked the email address in my commit to the github account

3. I added the mima backwards-excludes file

Thank you for your quick review and the feedback.

**Review by raboof — APPROVED** [review]

**Review by jrudolph — APPROVED** [review]
LGTM, thanks a lot, @nolledge.

**Comment by jrudolph** [comment]
>     1. I ran the HeaderParserBenchmark before and after my changes. I attached the results here:
> 
> 
> [header_benchmark_after.txt](https://github.com/akka/akka-http/files/5604858/header_benchmark_after.txt)
> [header_benchmark_before.txt](https://github.com/akka/akka-http/files/5604859/header_benchmark_before.txt)

Thanks for running that. Changes seem insignificant from a quick glance.

**Pull request merged**